applications/luci-app-cloudflared/htdocs/luci-static/resources/view/cloudflared/base.js

```diff
@@ -0,0 +1,109 @@
+/* This is free software, licensed under the Apache License, Version 2.0
+ *
+ * Copyright (C) 2024 Hilman Maulana <hilman0.0maulana@gmail.com>
+ */
+
+'use strict';
+'require form';
+'require poll';
+'require rpc';
+'require uci';
+'require view';
+
+var callServiceList = rpc.declare({
+	object: 'service',
+	method: 'list',
+	params: ['name'],
+	expect: { '': {} }
+});
+
+function getServiceStatus() {
+	return L.resolveDefault(callServiceList('cloudflared'), {}).then(function (res) {
+		var isRunning = false;
+		try {
+			isRunning = res['cloudflared']['instances']['cloudflared']['running'];
+		} catch (e) { }
+		return isRunning;
+	});
+}
+
+function renderStatus(isRunning) {
+	var spanTemp = '<label class="cbi-value-title">Status</label><div class="cbi-value-field"><em><span style="color:%s">%s</span></em></div>';
+	var renderHTML;
+	if (isRunning) {
+		renderHTML = String.format(spanTemp, 'green', _('Running'));
+	} else {
+		renderHTML = String.format(spanTemp, 'red', _('Not Running'));
+	}
+
+	return renderHTML;
+}
+
+return view.extend({
+	load: function() {
+		return Promise.all([
+			uci.load('cloudflared')
+		]);
+	},
+
+	render: function(data) {
+		var docsUrl = 'https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/configure-tunnels/tunnel-run-parameters/';
+		var m, s, o;
+
+		m = new form.Map('cloudflared', _('Cloudflared'),
+			_('Zero Trust Security services from Cloudflare to help you get maximum security both from outside and within the network.') + '<br />' +
+			_('Create and manage your cloudflared network on the <a %s>Cloudflare Zero Trust</a> website.')
+				.format('href="https://one.dash.cloudflare.com" target="_blank"') + '<br />' +
+			_('See <a %s>documentation</a> about the tunnel configuration parameters.')
+				.format('href="' + docsUrl + '" target="_blank"')
+		);
+
+		s = m.section(form.NamedSection, 'config', 'cloudflared');
+
+		o = s.option(form.DummyValue, 'service_status', _('Status'));
+		o.load = function () {
+			poll.add(function () {
+				return L.resolveDefault(getServiceStatus()).then(function (res) {
+					var view = document.getElementById('cbi-cloudflared-config-service_status');
+					view.innerHTML = renderStatus(res);
+				});
+			});
+		}
+		o.value = _('Collecting data...');
+
+		o = s.option(form.Flag, 'enabled', _('Enable'));
+		o.rmempty = false;
+
+		o = s.option(form.Value, 'token', _('Token'));
+		o.optional = true;
+		o.rmempty = false;
+
+		o = s.option(form.FileUpload, 'config', _('Config file'),
```

Comment:
It might make more sense to have a text edit field where you can paste in a simple yaml file, or assist the user in creating one, by filling a text field with the default parameters. The default provided yaml file needs editing anyway.

```yaml
cloudflared_binary: "cloudflared"
tunnel: "3d539f97-cd3a-4d8e-c33b-65e9099c7a8d"
credentials_file: "/Users/tunnel/.cloudflared/3d539f97-cd3a-4d8e-c33b-65e9099c7a8d.json"
origincert: "/Users/tunnel/.cloudflared/cert.pem"
ingress:
- hostname: named-tunnel-component-tests.example.com
  service: hello_world
- service: http_status:404
```

Have a look at https://openwrt.github.io/luci/jsapi/LuCI.form.TextValue.html and its `.default`. A `.validate` function could check for the essential keys to get cloudflared going (not essential). It will need a custom `write(section_id, formvalue)` function to write to the destination `/etc/cloudflared/config.yml` file. Simply:

```js
...
								fs.write('/etc/cloudflared/config.yml', formvalue);
...
```

Remember to add `'require fs';` at the start of this file.

Reply:
If the basic config editing is that needed then we may use this approach.
Generally speaking this is an old topic about text editor https://forum.openwrt.org/t/adding-a-file-browser-file-text-editor-and-tty-command-shell-for-luci-openwrt/138930
I hope one day we'll just have an option for the `form.FileUpload` like `o.editor = true`.

Anyway, a user should be able to configure everything with a GUI controls without need for raw editing. If something missing then probably we should improve the cloudflared package. Still if some rare option is missing it should be not a problem to upload a new config.

Reply:
So my proposition is to keep it as is and not over complicate for now. @systemcrash are you ok?